rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,204 @@
+            .thenAcceptAsync(capabiltities -> {
+                List<Registration> registrations = new LinkedList<>();
+                List<Unregistration> unregistrations = new LinkedList<>();
+                for (var entry : capabiltities) {
+                    var cap = entry.getLeft();
+                    var registration = entry.getRight();
+                    var method = cap.methodName();
+                    var existingRegistration = currentRegistrations.get(method);
+
+                    if (registration == null) {
+                        if (existingRegistration != null) {
+                            // this capability was removed
+                            logger.trace("{} is no longer supported by contributions", method);
+                            unregistrations.add(new Unregistration(cap.id(), cap.methodName()));
+                        }
+                        // nothing more to do
+                        continue;
+                    }
+
+                    if (existingRegistration != null) {
+                        if (Objects.deepEquals(registration.getRegisterOptions(), existingRegistration.getRegisterOptions())) {
+                            logger.trace("Options for {} did not change since last registration: {}", method, registration.getRegisterOptions());
+                            continue;
+                        }
+                        logger.trace("Options for {} changed since the previous registration; remove before adding again", method);
+                        unregistrations.add(new Unregistration(cap.id(), cap.methodName()));
+                    }
+
+                    logger.trace("Registering dynamic capability {}", method);
+                    registrations.add(registration);
+                }
+
+                if (!unregistrations.isEmpty()) {
+                    logger.debug("Unregistering dynamic capabilities: {}", unregistrations.stream().map(Unregistration::getMethod).collect(Collectors.toList()));
+                    try {
+                        client.unregisterCapability(new UnregistrationParams(unregistrations)).get();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                    } catch (ExecutionException e) {
+                        logger.error("Doing unregistrations failed!", e);
+                        return;
+                    }
+                    unregistrations.forEach(u -> currentRegistrations.remove(u.getMethod()));
+                } else {
+                    logger.debug("No unregistrations to do");
+                }
+                if (!registrations.isEmpty()) {
+                    logger.debug("Registering dynamic capabilities: {}", registrations.stream().map(Registration::getMethod).collect(Collectors.toList()));
+                    try {
+                        client.registerCapability(new RegistrationParams(registrations)).get();
```

Comment:
same here, we shouldn't do a `.get`  on a completable future inside of another completable future's body.